add implementation for Android PositioningService

johanvos opened the pull request.

johanvos committed:
add implementation for Android PositioningService

johanvos committed:
cleanup position code

jperedadnr commented on modules/position/src/main/java/com/gluonhq/attach/position/impl/AndroidPositionService.java:
```diff
@@ -0,0 +1,115 @@
+public class AndroidPositionService implements PositionService {
+
+    static {
+        System.loadLibrary("Position");
```
Should we check if we are in the JavaFX thread, like we are doing for other services?

johanvos replied:
No, this is not needed anymore as we create the JNIEnv in the native code. We just need to make sure we use the same thread here and also in the methods that invoke callbacks. But that is now checked in the native code.

jperedadnr commented on modules/position/src/main/java/com/gluonhq/attach/position/impl/AndroidPositionService.java:
```diff
@@ -0,0 +1,115 @@
+import com.gluonhq.attach.position.Parameters;
+import com.gluonhq.attach.position.Position;
+import com.gluonhq.attach.position.PositionService;
+import com.gluonhq.attach.util.Constants;
```
Unused import (probably it was used for debug log, which should be added at some point)

johanvos replied:
hard to detect those, but I removed a few

jperedadnr commented on modules/position/src/main/native/android/c/Position.c:
```diff
@@ -0,0 +1,133 @@
+
+// Dalvik handles
+jmethodID jPositionServiceStartMethod;
+jmethodID jPositionServiceStopMethod;
```
Maybe `jDalvikPositionServiceXXMethod`?
Or maybe we could use these prefixes: `jd`, `jg` for every class, methodID?

jperedadnr commented on modules/position/src/main/native/android/c/Position.c:
```diff
@@ -0,0 +1,133 @@
+    ATTACH_LOG_FINE("positionserviceobj: %p and init method = %p and class = %p\n", jPositionService, jPositionServiceInitMethod, jPositionServiceClass);
+    ATTACH_LOG_FINE("positionservice Starting: %p\n", jPositionService);
+    (*androidEnv)->CallVoidMethod(androidEnv, jPositionService, jPositionServiceStartMethod);
+    // (*androidVM)->DetachCurrentThread(androidVM);
```
Why we don't need this? Should we apply the same for other services?

johanvos replied:
my current approach is not to detach threads until an unload or so is called. We need to find a common pattern indeed.

jperedadnr commented on modules/position/src/main/native/android/c/Position.c:
```diff
@@ -0,0 +1,133 @@
+JNI_OnLoad_Position(JavaVM *vm, void *reserved)
+{
+#ifdef JNI_VERSION_1_8
+    graalVM = vm;
```
we do: `graalVM = getGraalVM()` in other services. But seems redundant?

johanvos replied:
yes, that is error-prone. we have graalVM defined as extern in graal.h and have to use it as such.

jperedadnr commented on modules/position/src/main/native/android/dalvik/DalvikPositionService.java:
```diff
@@ -0,0 +1,326 @@
+import android.location.LocationManager;
+import android.os.Bundle;
+import android.provider.Settings;
+import java.io.*;
+import java.util.*;
+import java.util.logging.*;
```
Avoid wildcards?

johanvos replied:
This is Dalvik code anyhow, I don't think we need coding conventions, but if you want, you can create a follow-up issue

jperedadnr commented on modules/position/src/main/native/android/dalvik/DalvikPositionService.java:
```diff
@@ -0,0 +1,326 @@
+
+    private static final String TAG = "GluonAttach";
+    private Activity activityContext;
+public static final Parameters DEFAULT_PARAMETERS = new Parameters(Parameters.Accuracy.MEDIUM, false);
```
`Parameters` is defined in GraalVM/Java side as common API, however to be able to use it from the Dalvik/Java side I see that it is duplicated (with different package name). Probably this is unavoidable?

johanvos replied:
can be discussed later

jperedadnr commented on modules/position/src/main/native/android/dalvik/DalvikPositionService.java:
```diff
@@ -0,0 +1,326 @@
+        }
+        gh = new EarthGravitationalModel();
+        try {
+            gh.load("egm180.nor");
```
This will fail for now, maybe add a TODO?

jperedadnr replied:
The file `egm180.not` is missing from this PR. It should be added to resources, so it is bundled by position:android.jar. However we don't have android/resources, so we'll need some tweak in our build, to avoid adding it to the other jars.

johanvos replied:
this is a follow-up issue: https://github.com/gluonhq/attach/issues/79

johanvos committed:
remove debug in Keyboard service

johanvos committed:
make sure to invoke dalvik methods on a dalvik Env

jperedadnr commented on modules/position/src/main/native/android/dalvik/AbstractMathTransform.java:
```diff
@@ -0,0 +1,222 @@
+
+package com.gluonhq.helloandroid;
+
+import java.io.Serializable;
```
These 3 classes for the altitude calculation don't require Android API. I was wondering if we should have them in the GraalVM/Java side, and invoke them from `AndroidPositionService` directly to modify the position readings.

johanvos replied:
sounds like a good follow-up issue. It would be better indeed probably. I didn't really look into those classes, but if those are simply to improve the accuracy of the altitude, +1.

johanvos committed:
don't detach current thread
Check if the thread is attached before invoking methods.
Don't share the JNIEnv with other services

johanvos committed:
process some comments
